include/oneapi/dpl/pstl/glue_algorithm_ranges_impl.h

```diff
@@ -314,6 +314,245 @@ struct __search_n_fn
+
+    std::ranges::copy_result<std::ranges::borrowed_iterator_t<_InRange>, std::ranges::borrowed_iterator_t<_OutRange>>
+    operator()(_ExecutionPolicy&& __exec, _InRange&& __in_r, _OutRange&& __out_r) const
+    {
+        const auto __dispatch_tag = oneapi::dpl::__ranges::__select_backend(__exec);
+
+        const auto _size = std::ranges::min(std::ranges::size(__in_r), std::ranges::size(__out_r));
+        oneapi::dpl::__internal::__ranges::__pattern_copy(__dispatch_tag, std::forward<_ExecutionPolicy>(__exec),
+            std::ranges::take_view(__in_r, _size), std::ranges::take_view(__out_r, _size));
+
+        return {std::ranges::begin(__in_r) + _size, std::ranges::begin(__out_r) +  _size};
+    }
+}; //__copy_fn
+}  //__internal
+
+inline constexpr __internal::__copy_fn copy;
+
+namespace __internal
+{
+struct __copy_if_fn
+{
+    template<typename _ExecutionPolicy, std::ranges::random_access_range _InRange,
+             std::ranges::random_access_range _OutRange, typename _Proj = std::identity,
+             std::indirect_unary_predicate<std::projected<std::ranges::iterator_t<_InRange>, _Proj>> _Pred>
+    requires oneapi::dpl::is_execution_policy_v<std::remove_cvref_t<_ExecutionPolicy>>
+        && std::ranges::sized_range<_InRange> && std::ranges::sized_range<_OutRange>
+        && std::indirectly_copyable<std::ranges::iterator_t<_InRange>, std::ranges::iterator_t<_OutRange>>
+    auto
+    operator()(_ExecutionPolicy&& __exec, _InRange&& __in_r, _OutRange&& __out_r, _Pred __pred, _Proj __proj = {}) const
+    {
+        const auto __dispatch_tag = oneapi::dpl::__ranges::__select_backend(__exec);
+        return oneapi::dpl::__internal::__ranges::__pattern_copy_if_ranges(__dispatch_tag, std::forward<_ExecutionPolicy>(__exec),
+            std::forward<_InRange>(__in_r), std::forward<_OutRange>(__out_r), __pred, __proj);
+    }
+}; //__copy_if_fn
+}  //__internal
+
+inline constexpr __internal::__copy_if_fn copy_if;
+
+namespace __internal
+{
+struct __merge_fn
+{
+    template<typename _ExecutionPolicy, std::ranges::random_access_range _R1, std::ranges::random_access_range _R2,
+             std::ranges::random_access_range _OutRange, typename _Comp = std::ranges::less,
+             typename _Proj1 = std::identity, typename _Proj2 = std::identity>
+    requires oneapi::dpl::is_execution_policy_v<std::remove_cvref_t<_ExecutionPolicy>> && std::ranges::sized_range<_R1>
+        && std::ranges::sized_range<_R2> && std::ranges::sized_range<_OutRange>
+        && std::mergeable<std::ranges::iterator_t<_R1>, std::ranges::iterator_t<_R2>,
+        std::ranges::iterator_t<_OutRange>, _Comp, _Proj1, _Proj2>
+    auto
+    operator()(_ExecutionPolicy&& __exec, _R1&& __r1, _R2&& __r2, _OutRange&& __out_r, _Comp __comp = {}, _Proj1 __proj1 = {},
+               _Proj2 __proj2 = {}) const
+    {
+        const auto __dispatch_tag = oneapi::dpl::__ranges::__select_backend(__exec);
```

Comment:
What's the behavior going to be if the `__out_r` is smaller than `__r1 + __r2`? Should we truncate `__r2` (and then `__r1` if it is larger than `__out_r`), or both `__r1` and `__r2` proportionally to fit `__out_r`?

Requesting `__out_r` to be sized at least as `__r1 + __r2` looks better to me. I can barely imagine that a user needs a "partial" merge, and it may be better to abort the execution rather than doing what is not expected.

Reply:
What's the behavior going to be if the __out_r is smaller than __r1 + __r2? Should we truncate __r2 (and then __r1 if it is larger than __out_r), or both __r1 and __r2 proportionally to fit __out_r?

Requesting __out_r to be sized at least as __r1 + __r2 looks better to me. I can barely imagine that a user needs a "partial" merge, and it may be better to abort the execution rather than doing what is not expected.

Reply:
We've discussed it offline and decided to investigate it later. I've added a todo comment.